visionmetrics/grounding/recall.py

```diff
@@ -9,6 +9,8 @@ class Recall(Metric):
 
     def __init__(self, iou_thresh=0.5, k=1):
         super().__init__()
+        if iou_thresh < 0 or iou_thresh > 1:
```

Comment:
There is a IOU implementation https://torchmetrics.readthedocs.io/en/stable/detection/intersection_over_union.html

Not quite sure if it is a good choice to re-use, or a test can be implemented using it?

Reply:
Not the same, IOU in torchmetrics computes the the IOU between a list of predictions and a list of boxes and return an average IOU,

Here we are computing the IOU between N's bbox and M's bbox, and return NxM matrix. 

```python
import torch
from torchmetrics.detection import IntersectionOverUnion
preds = [{"boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00],[300.00, 100.00, 315.00, 150.00], [100.00, 200.00, 315.00, 450.00]]), 
          "scores": torch.tensor([0.2, 0.3, 0.2]),
          "labels": torch.tensor([5, 6, 7])}]

target = [{"boxes": torch.tensor([[300.00, 100.00, 315.00, 150.00], [100.00, 200.00, 315.00, 450.00]]),"labels": torch.tensor([5, 7 ])}]
metric = IntersectionOverUnion()
metrics = metric(preds, target)
print(metrics)

# result: {'iou': tensor(0.3333)}

class test_iou():
    def __init__(self):
        pass

    def _box_area(self, boxes):
        assert boxes.ndim == 2 and boxes.shape[-1] == 4
        return (boxes[:, 2] - boxes[:, 0]) * (boxes[:, 3] - boxes[:, 1])

    def _box_iou(self, boxes1, boxes2):
        area1 = self._box_area(boxes1)
        area2 = self._box_area(boxes2)
        lt = torch.maximum(boxes1[:, None, :2], boxes2[:, :2])
        rb = torch.minimum(boxes1[:, None, 2:], boxes2[:, 2:])
        wh = (rb - lt).clip(min=0)
        inter = wh[:, :, 0] * wh[:, :, 1]
        union = area1[:, None] + area2 - inter
        iou = inter / union
        return iou

tester = test_iou()
iou = tester._box_iou(preds[0]['boxes'], target[0]['boxes'])
print(iou)

# result: tensor([[1., 0.], [1., 0.],[0., 1.]])
```